File: Algorithms/JavaScript/Easy/203_Remove_Linked_List_Elements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function(head, val) {
    if (head === null) {
        return null
    }

    var dummy = new ListNode(0)
    dummy.next = head

    var node = dummy
    while (node.next !== null) {
        var nextNode = node.next
        if (nextNode.val === val) {
            node.next = nextNode.next
        } else {
            node = nextNode
        }
    }

    return dummy.next
};

File: Algorithms/C++/Easy/203_Remove_Linked_List_Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode dummy(0);
        dummy.next = head;

        ListNode* node = &dummy;
        while (node->next != NULL) {
            if (node->next->val == val) {
                ListNode* tmp = node->next;
                node->next = node->next->next;
                delete tmp;
            } else {
                node = node->next;
            }
        }

        return dummy.next;
    }
};
